File: lib/java/src/main/java/uapi/internal/binary/PackMap.java
package uapi.internal.binary;

import static uapi.internal.binary.Pack.pack;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.msgpack.jackson.dataformat.MessagePackExtensionType;

public class PackMap {

    public static final byte UNDEFINED_BYTE = (byte) 18;

    static List<Object> packMap(Map<?, ?> m, List<Object> header,
            Map<Integer, BinaryPackNode> keyIndexMap) throws CannotPack {
        final var row = new ArrayList<Object>();
        for (final var entry : m.entrySet()) {
            if (entry.getKey() instanceof final String s) {
                throw new CannotPack();
            }

            final var key = (Integer) entry.getKey();
            final var keyIndex = keyIndexMap.get(key);

            final BinaryPackNode finalKeyIndex;
            if (keyIndex == null) {
                finalKeyIndex = new BinaryPackNode(header.size() - 1, new HashMap<>());

                if (entry.getValue() instanceof Map<?, ?>) {
                    header.add(new ArrayList<>(List.of(key)));
                } else {
                    header.add(key);
                }

                keyIndexMap.put(key, finalKeyIndex);
            } else {
                finalKeyIndex = keyIndex;
            }

            final Integer keyIndexValue = finalKeyIndex.value;
            final Map<Integer, BinaryPackNode> keyIndexNested = finalKeyIndex.nested;

            final Object packedValue;
            if (entry.getValue() instanceof Map<?, ?> m2) {
                final List<Object> nestedHeader;
                try {
                    nestedHeader = (List<Object>) header.get(keyIndexValue + 1);
                } catch (ClassCastException e) {
                    // No nesting available, so the data structure is inconsistent
                    throw new CannotPack();
                }

                packedValue = packMap(m2, nestedHeader, keyIndexNested);
            } else {
                if (header.get(keyIndexValue + 1) instanceof List) {
                    throw new CannotPack();
                }

                packedValue = pack(entry.getValue());
            }

            while (row.size() < keyIndexValue) {
                row.add(new MessagePackExtensionType(UNDEFINED_BYTE, new byte[0]));
            }

            if (row.size() == keyIndexValue) {
                row.add(packedValue);
            } else {
                row.set(keyIndexValue, packedValue);
            }
        }
        return row;
    }
}

File: lib/py/uapi/internal/binary/PackMap.py
from typing import Any, Dict, List, Tuple, Union
from msgpack import ExtType

from uapi.internal.binary.BinaryPackNode import BinaryPackNode
from uapi.internal.binary.CannotPack import CannotPack
from uapi.internal.binary.Pack import pack


UNDEFINED_BYTE = 18


def pack_map(m: Dict[Any, Any], header: List[Any], key_index_map: Dict[int, 'BinaryPackNode']) -> List[Any]:
    row: List[Any] = []
    for key, value in m.items():
        if isinstance(key, str):
            raise CannotPack()

        key = int(key)
        key_index = key_index_map.get(key)

        if key_index is None:
            final_key_index = BinaryPackNode(len(header) - 1, {})

            if isinstance(value, dict):
                header.append([key])
            else:
                header.append(key)

            key_index_map[key] = final_key_index
        else:
            final_key_index = key_index

        key_index_value = final_key_index.value
        key_index_nested = final_key_index.nested

        if isinstance(value, dict):
            try:
                nested_header = header[key_index_value + 1]
                if not isinstance(nested_header, list):
                    raise TypeError()
            except (IndexError, TypeError):
                raise CannotPack()

            packed_value = pack_map(value, nested_header, key_index_nested)
        else:
            if isinstance(header[key_index_value + 1], list):
                raise CannotPack()

            packed_value = pack(value)

        while len(row) < key_index_value:
            row.append(ExtType(UNDEFINED_BYTE, b''))

        if len(row) == key_index_value:
            row.append(packed_value)
        else:
            row[key_index_value] = packed_value

    return row
